File: eventweave/runtime/sinks/stdout.py
"""Stdout sink that prints events as JSON lines."""

from __future__ import annotations

import json

from eventweave.core.event import Event
from eventweave.encoders.base import Encoder
from eventweave.runtime.sink import Sink


class StdoutSink(Sink):
    """Print each event to stdout as JSON or via an encoder."""

    def __init__(self, encoder: Encoder | None = None) -> None:
        self.encoder = encoder
        self._count = 0
        self._failed = 0

    def open(self) -> None:
        pass

    def write(self, event: Event) -> None:
        if self.encoder is not None:
            result = self.encoder.encode(event)
            if not result.success:
                self._failed += 1
                raise RuntimeError(f"encode failed: {result.error_reason}")
            print(result.output)
        else:
            print(json.dumps(event.model_dump(), default=str, ensure_ascii=False))
        self._count += 1

    def close(self) -> None:
        pass

    def flush(self) -> None:
        pass

    def count(self) -> int:
        return self._count

    def failed(self) -> int:
        return self._failed

File: runtime-go/internal/sinks/stdout/stdout.go
package stdout

import (
	"encoding/json"
	"fmt"

	"github.com/linxinhong/eventweave/runtime-go/internal/encoder"
	"github.com/linxinhong/eventweave/runtime-go/internal/event"
)

// Sink prints events as JSON lines or encoded lines.
type Sink struct {
	enc    encoder.Encoder
	count  int
	failed int
}

// New creates a stdout sink.
func New(enc ...encoder.Encoder) *Sink {
	s := &Sink{}
	if len(enc) > 0 {
		s.enc = enc[0]
	}
	return s
}

// Open initializes the sink.
func (s *Sink) Open() error { return nil }

// Write prints one event.
func (s *Sink) Write(ev event.Event) error {
	var data []byte
	var err error
	if s.enc != nil {
		data, err = s.enc.Encode(ev)
	} else {
		data, err = json.Marshal(ev)
	}
	if err != nil {
		s.failed++
		return err
	}
	fmt.Println(string(data))
	s.count++
	return nil
}

// Flush is a no-op for stdout.
func (s *Sink) Flush() error { return nil }

// Close is a no-op for stdout.
func (s *Sink) Close() error { return nil }

// Count returns emitted events.
func (s *Sink) Count() int { return s.count }

// Failed returns failed writes.
func (s *Sink) Failed() int { return s.failed }
